### PR TITLE
fix `===` when encountering null pointer

### LINKED ISSUE #44712: Regression in `===` (caused by #43658?)
Running the following:

```jl
using TensorKit, TensorKitManifolds, StableRNGs
V = ℂ^4
T = Float64
rng = StableRNG(0)
W, = leftorth(TensorMap((x...)-> randn(rng, x...), T, V*V*V, V*V); alg = Polar())
X = TensorMap((x...)-> randn(rng, x...), T, space(W))
Y = TensorMap((x...)-> randn(rng, x...), T, space(W))
Δ = Unitary.project(X, W)
Θ = Unitary.project(Y, W)

Δ === Θ
```

we get `false`on 1.7, while on 1.8 this returns `true` 

Here is a proof they are not `===`:

```jl
julia> Θ.A.data |> sum
1.1102230246251565e-15

julia> Δ.A.data |> sum
-4.6074255521944e-15
```

Reverting https://github.com/JuliaLang/julia/pull/43658 makes it go away so I guess the bug is from there.

cc @NHDaly

### PR DESCRIPTION
**Pull request by Moelf**
closes #44712

### CODE REVIEW

File: src/builtins.c [hunk 1]
```diff
@@ -94,20 +94,22 @@ static int NOINLINE compare_fields(const jl_value_t *a, const jl_value_t *b, jl_
...
+                // we need to check for undef first since undef struct
                 // may have fields that are different but should still be treated as equal.
-                jl_value_t *ptra = ((jl_value_t**)ao)[ft->layout->first_ptr];
-                jl_value_t *ptrb = ((jl_value_t**)bo)[ft->layout->first_ptr];
-                if (ptra == NULL && ptrb == NULL) {
-                    return 1;
+                int32_t idx = ft->layout->first_ptr;
+                jl_value_t *ptra = ((jl_value_t**)ao)[idx];
+                jl_value_t *ptrb = ((jl_value_t**)bo)[idx];
+                if ((ptra != NULL && ptrb == NULL) || (ptra == NULL && ptrb != NULL)) {
```
Comment (vtjnash) [inline src/builtins.c:111]: Can be simplified to:
```suggestion
                if ((ptra == NULL) != (ptrb == NULL)) {
```

**Comment by KristofferC** [comment]
Possible to make a test for this since it is clearly something that is lacking?

**Comment by vtjnash** [comment]
```julia
julia> struct AB
       a::Some{Any}
       b::Int
       AB() = new()
       end

julia> unsafe_load(Ptr{AB}(pointer(Int[0,1]))) === unsafe_load(Ptr{AB}(pointer(Int[0,2])))
true

julia> unsafe_load(Ptr{AB}(pointer(Int[0,1]))), unsafe_load(Ptr{AB}(pointer(Int[0,2])))
(AB(#undef, 1), AB(#undef, 2))
```

**Comment by Moelf** [comment]
that returns `true` even in 1.7...

**Comment by Moelf** [comment]
Error seems unrelated

File: test/core.jl [hunk 2]
```diff
@@ -92,6 +92,14 @@ let abcd = ABCDconst(1, 2, 3, 4)
...
 end
 
+# test `===` handling null pointer in struct #44712
+struct N44712
+    a::Some{Any}
+    b::Int
+    AB() = new()
+end
+
+@test unsafe_load(Ptr{N44712}(pointer(Int[0,1]))) !== unsafe_load(Ptr{N44712}(pointer(Int[0,2])))
```
Comment (vtjnash) [inline test/core.jl:102]: our unsafe test might need a little less danger
```suggestion
    N44712() = new()
end
let a  = Int[0, 1], b = Int[0, 2]
    GC.@preserve a b begin
        @test unsafe_load(Ptr{N44712}(pointer(a))) !== unsafe_load(Ptr{N44712}(pointer(b)))
    end
end
```

**Comment by KristofferC** [comment]
I rebased it to see if that makes CI happier.

**Comment by KristofferC** [comment]
Looks like there is still a bug here:

```jl
julia> begin
       Experimental.@force_compile
       @test !platforms_match(P("x86_64", "linux"; julia_version=v"1.5.0"),
                              P("x86_64", "linux"; julia_version=v"1.6.0"))
       end
Test Failed at REPL[20]:3
  Expression: !(platforms_match(P("x86_64", "linux"; julia_version = v"1.5.0"), P("x86_64", "linux"; julia_version = v"1.6.0")))
ERROR: There was an error during testing

julia> begin
       #Experimental.@force_compile
       @test !platforms_match(P("x86_64", "linux"; julia_version=v"1.5.0"),
                              P("x86_64", "linux"; julia_version=v"1.6.0"))
       end
Test Passed
```

**Comment by KristofferC** [comment]
Presumably in some way since it doesn't happen on e.g the master branch.

**Comment by Moelf** [comment]
```julia
julia> begin
           Experimental.@force_compile
           p2 = Platform("x86_64", "linux"; julia_version=v"1.6.0")
           @show tags(p2)["julia_version"]
       end
(tags(p2))["julia_version"] = "1.6.0"
"1.6.0"
```


eh, I don't understand

File: src/builtins.c [hunk 3]
```diff
@@ -94,20 +94,25 @@ static int NOINLINE compare_fields(const jl_value_t *a, const jl_value_t *b, jl_
...
-                    return 1;
+                int32_t idx = ft->layout->first_ptr;
+                jl_value_t *ptra = ((jl_value_t**)ao)[idx];
+                jl_value_t *ptrb = ((jl_value_t**)bo)[idx];
+                if ((ptra == NULL) != (ptrb == NULL)) {
+                    return 0;
+                }
+                else {
+                    continue;
                 }
```
Comment (vtjnash) [inline src/builtins.c:116]:
```suggestion
                else if (ptra == NULL) { // implies ptrb == NULL
                    continue; // skip this field (it is #undef)
                }
```

Comment (Moelf) [reply src/builtins.c:116]: Looks like we don't have enough tests... My bad

**Comment by KristofferC** [comment]
Would be nice to have a dedicated test for the latest found problem. Maybe you can cook up one more @vtjnash?

**Comment by Moelf** [comment]
I think I can cook one up today

**Comment by Moelf** [comment]
it's harder than I thought, I try to make two immutable structs both containing non-null pointers but are different, I tried something like this but it segfaults:
```
julia> let a  = Int[1, 1], b = Int[2, 1]
           GC.@preserve a b begin
               unsafe_load(Ptr{N44712}(pointer(a))) === unsafe_load(Ptr{N44712}(pointer(b)))
           end
       end
```

**Comment by vtjnash** [comment]
neither of those contain a null pointer

**Comment by vtjnash** [comment]
What you were breaking is the much more trivial case of some pointer followed by some value:
```julia
(("", 0),) === (("", 1),)
```
normally we compile this, so we don't observe often if it broke

**Comment by Moelf** [comment]
The windows test failure is unrelated (I checked the CI a bunch of other PRs test also failing)

**Pull request merged**